File: src/main/java/org/appsugar/archetypes/security/UserDetailServiceImpl.java
package org.appsugar.archetypes.security;

import lombok.Getter;
import lombok.Setter;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;
import org.appsugar.archetypes.domain.User;
import org.appsugar.archetypes.repository.UserRepository;
import org.appsugar.archetypes.system.Permissions;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.context.properties.ConfigurationProperties;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import javax.annotation.PostConstruct;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * spring security 用户信息查找
 *
 * @author shenliuyang
 * @version 1.0.0
 * @package org.appsugar.archetypes.service
 * @className UserDetailServiceImpl
 * @date 2021-07-06  19:31
 */
@Service
@Slf4j
@ConfigurationProperties("management.security")
public class UserDetailServiceImpl implements UserDetailsService {
    @Autowired
    UserRepository userRepository;
    @Autowired
    PasswordEncoder encoder;
    @Getter
    @Setter
    String username;
    @Getter
    @Setter
    String password;

    UserDetails randomUserDetails;

    @Override
    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
        //排除内部系统访问 spring boot admin server
        if (StringUtils.equals(username, getUsername())) {
            return randomUserDetails;
        }
        User user = userRepository.findOneByLoginName(username);
        if (Objects.isNull(user)) {
            log.info("user  login failed because  {} do not exist", username);
            throw new UsernameNotFoundException(username);
        }
        String loginName = username;
        String pwd = encoder.encode(user.getPassword());
        String split = User.PERMISSION_SPLIT_CHAR;
        Set<String> permissions = user.getRoles().stream().flatMap(e -> Arrays.stream(e.getPermissions().split(","))).collect(Collectors.toSet());
        for (String s : user.getPermissions().split(split)) {
            permissions.add(s);
        }
        return new LoginUser(user.getId(), loginName, pwd, permissions, Permissions.modifyCount, System.currentTimeMillis());
    }

    @PostConstruct
    public void postConstruct() {
        this.randomUserDetails = new org.springframework.security.core.userdetails.User(username, encoder.encode(password), new HashSet<>());
    }
}

File: src/main/kotlin/org/appsugar/archetypes/service/UserDetailServiceImpl.kt
package org.appsugar.archetypes.service

import org.appsugar.archetypes.extension.getLogger
import org.appsugar.archetypes.repository.UserRepository
import org.appsugar.archetypes.web.Permission
import org.springframework.boot.context.properties.ConfigurationProperties
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Primary
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.User
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import java.io.Serializable

@Service
@Primary
@ConfigurationProperties("spring.security.user")
class UserDetailServiceImpl(val userRepository: UserRepository, val passwordEncoder: PasswordEncoder) : UserDetailsService {

    var name = "user"
    var password = System.currentTimeMillis().toString()

    companion object {
        var endpointPermission = "endpoint"
    }

    val logger = getLogger<UserDetailServiceImpl>()

    override fun loadUserByUsername(username: String): UserDetails {
        if (username == name) return UserPrincipal(0, name, passwordEncoder.encode(password), mutableListOf(SimpleGrantedAuthority(endpointPermission)))
        val user = userRepository.findByLoginName(username)
                ?: throw UsernameNotFoundException("user $username not found")
        var permissions = mutableSetOf<String>().apply { user.permissions }
        for (role in user.roles) {
            permissions.addAll(role.permissions)
        }
        if (permissions.contains("*")) permissions.addAll(Permission.values().map { it.value })
        return UserPrincipal(user.id, user.loginName, passwordEncoder.encode(user.password), permissions.map { SimpleGrantedAuthority(it) }.toMutableList())
    }


}

@Configuration
class PasswordEncoderConfiguration {
    @Bean
    fun encoder() = BCryptPasswordEncoder(11)
}

class UserPrincipal<T : GrantedAuthority>(val id: Long, username: String, password: String, authorities: MutableCollection<T>)
    : User(username, password, authorities), Serializable {
    val attributes = mutableMapOf<String, Any>()
}